File: 679. 24 Game/Java/Solution.java
class Solution {
    public boolean judgePoint24(int[] nums) {
        class Utils {
            double EPS = 0.001;

            boolean backTrace(List<Double> nums) {
                if (nums.size() == 1) {
                    return Math.abs(nums.get(0) - 24) < EPS;
                }
                for (int i = 0; i < nums.size(); i++) {
                    double a = nums.get(i);
                    for (int j = 0; j < i; j++) {
                        double b = nums.get(j);
                        List<Double> candidates = new ArrayList<>();
                        candidates.add(a + b);
                        candidates.add(a - b);
                        candidates.add(b - a);
                        candidates.add(a * b);
                        if (Math.abs(a) > EPS) {
                            candidates.add(b / a);
                        }
                        if (Math.abs(b) > EPS) {
                            candidates.add(a / b);
                        }
                        nums.remove(i);
                        nums.remove(j);
                        for (double candidate : candidates) {
                            nums.add(candidate);
                            if (backTrace(nums)) {
                                return true;
                            }
                            nums.remove(nums.size() - 1);
                        }
                        nums.add(j, b);
                        nums.add(i, a);
                    }
                }
                return false;
            }
        }
        Utils utils = new Utils();
        List<Double> temp = new ArrayList<>();
        for (int num : nums) {
            temp.add((double)num);
        }
        return utils.backTrace(temp);
    }
}

File: 679. 24 Game/C++/Solution.cpp
class Solution {
public:
    bool judgePoint24(vector<int> &nums) {
        class Utils {
            double EPS = 0.001;
        public:
            bool backTrace(vector<double> &nums) {
                if (nums.size() == 1) {
                    return abs(nums[0] - 24) < EPS;
                }
                for (int i = 0; i < nums.size(); i++) {
                    double a = nums[i];
                    for (int j = 0; j < i; j++) {
                        double b = nums[j];
                        vector<double> candidates = {a + b, a - b, b - a, a * b};
                        if (abs(a) > EPS) {
                            candidates.push_back(b / a);
                        }
                        if (abs(b) > EPS) {
                            candidates.push_back(a / b);
                        }
                        nums.erase(nums.begin() + i);
                        nums.erase(nums.begin() + j);
                        for (auto candidate : candidates) {
                            nums.push_back(candidate);
                            if (backTrace(nums)) {
                                return true;
                            }
                            nums.pop_back();
                        }
                        nums.insert(nums.begin() + j, b);
                        nums.insert(nums.begin() + i, a);
                    }
                }
                return false;
            }
        };
        Utils utils;
        vector<double> temp;
        for (int num : nums) {
            temp.push_back((double) num);
        }
        return utils.backTrace(temp);
    }
};